### PR TITLE
User input

### PR DESCRIPTION
**Pull request by riamon-v**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,3 +30,13 @@
 *.exe
 *.out
 *.app
+
+#AutoGenerate files
+*~
+#*#
```
Comment (Gaspard--) [inline .gitignore:36]: I'll let @Arignir explain why this shouldn't be your approach.

File: source/main.cpp [hunk 2]
```diff
@@ -1,8 +1,43 @@
...
-int main()
-{
-  display::Display<display::WaylandSurface> display;
+	UserInput::get().egl_display = eglGetDisplay(UserInput::get().display);
+	eglInitialize(UserInput::get().egl_display, NULL, NULL);
+
+	UserInput::get().xkb_context = xkb_context_new (XKB_CONTEXT_NO_FLAGS);
+
+	UserInput::createWindow();
+
```
Comment (Gaspard--) [inline source/main.cpp:28]: All of lines above dont look clean. Maybe should be behind an init method or in a constructor? Overall I don't see why a singleton is necessary.

File: source/main.cpp [hunk 3]
```diff
@@ -1,8 +1,43 @@
...
 
-int main()
-{
-  display::Display<display::WaylandSurface> display;
+	UserInput::get().egl_display = eglGetDisplay(UserInput::get().display);
+	eglInitialize(UserInput::get().egl_display, NULL, NULL);
+
+	UserInput::get().xkb_context = xkb_context_new (XKB_CONTEXT_NO_FLAGS);
+
+	UserInput::createWindow();
```
Comment (Gaspard--) [inline source/main.cpp:27]: create usually returns a object or something like that. This seems counter intuitive.

File: source/main.cpp [hunk 4]
```diff
@@ -1,8 +1,43 @@
...
+
+	while (UserInput::get().running) {
+		wl_display_dispatch_pending(UserInput::get().display);
+		UserInput::get().getWindow().draw();
+	}
+
+//	delete win;
+	eglTerminate(UserInput::get().egl_display);
```
Comment (Gaspard--) [inline source/main.cpp:35]: This screams of missing RAII.

File: source/user-input/Listeners.cpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+Listeners::Listeners() {
+  const PointerListener pL = PointerListener(listeners);
+  const KeyboardListener kL = KeyboardListener(listeners);
+  const SeatListener sL = SeatListener(listeners);
+  const RegistryListener rL = RegistryListener(listeners);
+  const WindowListener wL = WindowListener(listeners);
+}
+
+Listeners::~Listeners() {
+
```
Comment (Gaspard--) [inline source/user-input/Listeners.cpp:12]: Empty destructor is rarely a good choice. `default` it or something.

File: source/user-input/UserInput.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+WindowHandler &UserInput::getWindow() {
+	return *win;
+}
+
+void UserInput::createWindow() {
+	UserInput::instance.win = new WindowHandler(UserInput::instance.egl_display,
+																							UserInput::instance.shell,
+																							UserInput::instance.compositor);
```
Comment (Gaspard--) [inline source/user-input/UserInput.cpp:31]: This indentation is unreadable.

File: source/user-input/listeners/KeyboardListener.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	close(fd);
+	xkb_state_unref(UserInput::get().xkb_state);
+	UserInput::get().xkb_state = xkb_state_new(UserInput::get().keymap);
+}
+
+void KeyboardListener::keyboard_enter (void *data, struct wl_keyboard *keyboard, uint32_t serial, struct wl_surface *surface, struct wl_array *keys) {
+  printf("Entering the window\n");
+}
+
+void KeyboardListener::keyboard_leave (void *data, struct wl_keyboard *keyboard, uint32_t serial, struct wl_surface *surface) {
```
Comment (Gaspard--) [inline source/user-input/listeners/KeyboardListener.cpp:29]: You seem to not be compiling with `-Wall`. Unused params shouldn't be named.

Comment (Gaspard--) [reply source/user-input/listeners/KeyboardListener.cpp:29]: This concerns a bunch of functions. I won't list them.

File: source/user-input/listeners/RegistryListener.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#include "user-input/RegistryListener.hpp"
+
+RegistryListener::RegistryListener(std::map<std::string, Listener> &listeners) {
+  static struct wl_registry_listener registry_listener =
```
Comment (Gaspard--) [inline source/user-input/listeners/RegistryListener.cpp:4]: What about `constexpr` ?

File: source/user-input/listeners/RegistryListener.cpp [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+          {&RegistryListener::registry_add_object,
+           &RegistryListener::registry_remove_object};
+  listeners.emplace("registry", registry_listener);
+}
+
+void RegistryListener::registry_add_object (void *data, struct wl_registry *registry, uint32_t name, const char *interface, uint32_t version) {
+	if (!strcmp(interface,"wl_compositor")) {
```
Comment (Gaspard--) [inline source/user-input/listeners/RegistryListener.cpp:11]: Please use `std::strcmp`.

File: include/user-input/UserInput.hpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <wayland-client.h>
+#include <wayland-egl.h>
+#include <EGL/egl.h>
+#include <GL/gl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <xkbcommon/xkbcommon.h>
+#include <string.h>
```
Comment (Gaspard--) [inline include/user-input/UserInput.hpp:10]: use `<cstring>` instead.

File: include/user-input/UserInput.hpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+#include <stdio.h>
```
Comment (Gaspard--) [inline include/user-input/UserInput.hpp:11]: use `<cstdio>` instead.

**Review by Gaspard-- — CHANGES_REQUESTED** [review]
Overall the code isn't great, but isn't bad either. I'd be ok with merging this, but it could be quite a bit cleaner and it would be interesting to investigate if a singleton is really necessary.

File: .gitignore (hunk 1 above)
Comment (Arignir) [reply .gitignore:36]: At least, these ones i don't care, they are shared across so many editors that it's fine.

But in other cases, try to put personal files in `/.git/info/exclude` and not `/.gitignore`

Comment (Gaspard--) [reply .gitignore:36]: Well a little lower, there's also a random directory (I know, `build/` is not random, but only to people who name it "build") which definitely shouldn't be here.

**Comment by Zeng-Zer** [comment]
Seems like nobody is working on this PR, maybe we should close it ?

**Comment by Gaspard--** [comment]
Yes

**Pull request closed without merging**